Add mTLS hello world example

Ryanfsdf opened the pull request:
This is the rough implementation of a hello world example which uses mTLS. 

This isn't working right now - I think the issue is that the server side certificate pool isn't set up properly. I'll need to play around with openssl a bit more to figure out how to get it to work.

I just wanted to get this looked at to make sure I'm headed in the right direction.

Ryanfsdf committed:
Add mTLS hello world example

matthewstevenson88 commented:
Thanks for the PR Ryan! This is definitely the right idea.

3 general comments:

1. Please configure the client and server to use TLS 1.3. (At the moment, they are configured to negotiate a TLS version without a specified minimum. So if the server was malicious, it could select e.g. TLS 1.0 or SSL 3.0, which are broken.)

2. For the client's cert/key, please pass the path as a flag. You will also need to pass the root cert as a flag and use the root cert to initialize the cert pool. You did this correctly on the server-side, but we also need it on the client-side since we're doing mTLS rather than vanilla TLS.  (For this example, the cert for the client and the root cert can be the same.)

3. Please also change the address of the server so that it is passed by a flag.

Ryanfsdf committed:
Parameterize key locations and address

Ryanfsdf commented:
Thanks for the review! I addressed the review comments. I left the port number unparameterized but I've parameterized the rest of the arguments. Would keeping them as command line arguments be okay? Or would we eventually have to move over to something like gflags? 

PTAL!

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
```
s/2015/2020

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+	"google.golang.org/grpc"
+	pb "google.golang.org/grpc/security/s2a/examples/helloworld"
+)
+
```
Please define variables for each of the command-line flags. For example,

`var {
  serverAddr = flags.String(...)
}`

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+const (
+	port        = "50051"
```
Please make the full server address into a command-line flag, rather than just using the port number.

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+func main() {
+	if len(os.Args) != 5 {
```
Please use the "flags" module rather than "os". See: https://gobyexample.com/command-line-flags

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+
+const (
+	port        = "50051"
+	defaultName = "world"
```
Is `defaultName` used for anything?

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+		keyFile,
+	)
+	if err != nil {
+		log.Fatalf("Failed to setup TLS certificate: %v", err)
```
Can you make this slightly more specific? For example, "Failed to load client's X509 certificate.".

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,93 @@
+	config := &tls.Config{
+		Certificates: []tls.Certificate{certificate},
+		RootCAs:      certPool,
+		MinVersion:   tls.VersionTLS13,
```
Please also set the max TLS version. (TLS 1.3 is the max right now, but it's best to set the max to be 1.3 so nothing changes unexpectedly in the future.)

matthewstevenson88 commented on security/s2a/examples/greeter_server/main.go:
```diff
@@ -0,0 +1,97 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
```
s/2015/2020

matthewstevenson88 commented on security/s2a/examples/greeter_server/main.go:
```diff
@@ -0,0 +1,97 @@
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
```
Same comment as for client: please use "flags" rather "os".

matthewstevenson88 commented on security/s2a/examples/helloworld/helloworld.proto:
```diff
@@ -0,0 +1,38 @@
+syntax = "proto3";
+
+option go_package = "google.golang.org/grpc/examples/helloworld/helloworld";
+option java_multiple_files = true;
```
Please omit the java options.

Ryanfsdf replied:
Do you think this is worth cleaning up in the non-mtls example as well? The proto file was taken from that example.

matthewstevenson88 replied:
No, that's ok. That proto is owned by gRPC and not us, so we can leave it.

matthewstevenson88 commented on security/s2a/examples/helloworld/helloworld.proto:
```diff
@@ -0,0 +1,38 @@
+
+syntax = "proto3";
+
+option go_package = "google.golang.org/grpc/examples/helloworld/helloworld";
```
Can you change this to a security/s2a package?

matthewstevenson88 commented on security/s2a/examples/helloworld/helloworld.proto:
```diff
@@ -0,0 +1,38 @@
+// Copyright 2015 gRPC authors.
```
s/2015/2020

matthewstevenson88 commented:
Thanks for the PR Ryan! I've left a bunch of comments. Additionally, please add the certificates and keys that you used to a `security/s2a/testdata` directory.

Please ping me when it's ready for another pass! :)

matthewstevenson88 commented:
One other comment: please use `gofmt` or similar to format the code. See:

https://blog.golang.org/gofmt

Ryanfsdf committed:
Address review comments and add test data

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+func main() {
+	serverAddr := flag.String("server address", "localhost", "address of the server")
```
Please change the flag name to "server_address".

matthewstevenson88 replied:
Also, the flag should be the full address of the server, e.g. "localhost:50051".

Ryanfsdf committed:
Fix flag name parsing

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+
+func main() {
+	serverAddr := flag.String("server address", "localhost", "address of the server")
+	port := flag.String("port", "50051", "port number to use for connection")
```
Please remove the port flag.

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+func main() {
+	serverAddr := flag.String("server address", "localhost", "address of the server")
+	port := flag.String("port", "50051", "port number to use for connection")
+	rootCert := flag.String("root certificate", "../../testdata/ca.cert", "path to root X509 certificate")
```
Please rename this flag "server_root_cert_pem_path".

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+	serverAddr := flag.String("server address", "localhost", "address of the server")
+	port := flag.String("port", "50051", "port number to use for connection")
+	rootCert := flag.String("root certificate", "../../testdata/ca.cert", "path to root X509 certificate")
+	certFile := flag.String("client certificate", "../../testdata/client.pem", "path to client's X509 certificate")
```
Please rename this flag "client_cert_pem_path".

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+	port := flag.String("port", "50051", "port number to use for connection")
+	rootCert := flag.String("root certificate", "../../testdata/ca.cert", "path to root X509 certificate")
+	certFile := flag.String("client certificate", "../../testdata/client.pem", "path to client's X509 certificate")
+	keyFile := flag.String("client private key", "../../testdata/client.key", "path to client's private key")
```
Please rename this flag "client_key_pem_path".

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// Set up a connection to the server.
+	fullServerAddr := *serverAddr + ":" + *port
```
Based on the changes to the flags, this should just be `*serverAddr`.

matthewstevenson88 commented on security/s2a/examples/greeter_client/main.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+var (
+	serverAddr string
```
Please move the declaration of the flags to here; that is, have

`serverAddr = flag.String("server_address", "localhost:50051", "The address of the gRPC server.")`

matthewstevenson88 commented on security/s2a/examples/greeter_server/main.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+var (
+	port     string
```
Please make the same changes to the flags that I mentioned in the client file.

matthewstevenson88 commented on security/s2a/testdata/ca.cert:
```diff
@@ -0,0 +1,30 @@
+-----BEGIN CERTIFICATE-----
```
Can you add a `README.md` to this folder that explains what commands you used to generate all the certs/keys? This way they can be reproduced.

matthewstevenson88 commented:
Thanks Ryan! I've left a few more comments, mostly related to the format of the flags that we discussed offline.

Ryanfsdf committed:
Update flag names and add readme

Ryanfsdf commented:
Thanks for the review! Addressed all the review comments and also verified that the client/server and flag parsing works with the following:
1) run `go run main.go` in the server folder
2) run `go run main.go` in the client folder
3) verify that the correct message is printed

Then verified that passing the flag works
1) run `go run main.go -port=10010` in the server folder
2) run `go run main.go` in the client folder. Nothing should be printed because the port numbers won't match up
3) run `go run main.go -server_address=localhost:10010` in the client folder. The correct message should be printed now because the port numbers match up.

matthewstevenson88 commented on security/s2a/examples/greeter_server/main.go:
```diff
@@ -0,0 +1,96 @@
+	return &pb.HelloReply{Message: "Hello " + in.GetName()}, nil
+}
+
+var (
```
Please move this to right after the import block.

Ryanfsdf replied:
Done

matthewstevenson88 reviewed (approved):
Thanks Ryan! I left one small comment and I'm approving. Also adding Cesar as a reviewer to approve for readability.

Once Cesar has approved, please squash and merge.

Ryanfsdf committed:
Move var declaration block

cesarghali commented on security/s2a/testdata/README.md:
```diff
@@ -0,0 +1,19 @@
+## Key Generation
```
I think we can restructure this section as follows:
1. CA key and cert generation.
2. Client key and cert generation.
3. Server key and cert generation.

Also, please the exact command that you used to generate the files. For example, use the exact file name that you used instead of `<FILE_NAME>`. This way the instructions can directly reflect the files that you are adding in this directory.

Ryanfsdf committed:
Address review comments

Ryanfsdf commented:
@cesarghali Thank you for the review! I addressed all the review comments, PTAL.

cesarghali commented on security/s2a/testdata/README.md:
````diff
@@ -1,19 +1,19 @@
+cp ca.key client.key
+openssl req -new -x509 -key client.key -sha256 -subj "/C=US/ST=NJ/O=CA, Inc." -days 365 -out client.pem
 ```
+Note that the same private key was used for the client and ca.
````
Any specific reason why we're doing this? If the same key is used, then the client is basically the CA :)

Can we chain the client cert from the CA cert the same we do for the server? We would then have similar commands for generating client and server certs. The difference would be the filenames.

Ryanfsdf replied:
No specific reason, I just couldn't figure out how to chain the client cert from the CA cert (I'm not very familiar with openssl). I didn't realize I could just do the same as what I did for the server.

I updated the PR with the new instructions and client key/pem pair.

Ryanfsdf committed:
Update client key and pem

cesarghali reviewed (approved):
Go approval. Thanks for the PR!

Ryanfsdf committed:
Update client and server pem

Ryanfsdf commented:
Updated the client and server certificates to include localhost in the subject. This is necessary for us to reuse the keys/certs for the C++ example, since the C++ side requires that localhost is in the subject. Also updated the README to account for this change.

matthewstevenson88 commented:
LGTM. @Ryanfsdf please squash and merge this PR, and in the future add new commits as separate PRs. Thanks! :)
